File: core/aggregators/utils/bluesky.py
"""
Bluesky utilities for inline embed rendering.

Provides functions for:
- Detecting Bluesky URLs
- Extracting post info (handle/DID + record key)
- Resolving handles to DIDs
- Fetching post data from the public Bluesky API
- Extracting images from posts
- Building rich HTML embeds
"""

import html
import logging
import re
from typing import Any, Dict, List, Optional, Tuple

import requests

from .block_parser import is_safe_url

logger = logging.getLogger(__name__)

# Public (unauthenticated) Bluesky AppView API endpoint
BSKY_API_BASE = "https://public.api.bsky.app"


def is_bluesky_url(url: str) -> bool:
    """
    Check if a URL is a Bluesky URL.

    Handles:
    - bsky.app
    - staging.bsky.app

    Args:
        url: URL to check

    Returns:
        True if URL is from Bluesky
    """
    if not url:
        return False

    return "bsky.app" in url


def extract_bluesky_post_info(url: str) -> Optional[Tuple[str, str]]:
    """
    Extract the actor (handle or DID) and record key from a Bluesky post URL.

    Pattern: /profile/{handle_or_did}/post/{rkey}

    Args:
        url: Bluesky post URL

    Returns:
        Tuple of (actor, rkey) if found, None otherwise
    """
    if not url:
        return None

    match = re.search(r"/profile/([^/]+)/post/([^/?#]+)", url)
    if match:
        return match.group(1), match.group(2)

    return None


def resolve_bluesky_did(actor: str, timeout: int = 10) -> Optional[str]:
    """
    Resolve a Bluesky handle to a DID.

    If the actor is already a DID (starts with "did:"), it is returned as-is.

    Args:
        actor: Bluesky handle (e.g. "user.bsky.social") or DID
        timeout: Request timeout in seconds

    Returns:
        DID if resolved, None if failed
    """
    if not actor:
        return None

    if actor.startswith("did:"):
        return actor

    try:
        url = f"{BSKY_API_BASE}/xrpc/com.atproto.identity.resolveHandle"
        headers = {"User-Agent": "Yana/1.0"}

        response = requests.get(url, params={"handle": actor}, headers=headers, timeout=timeout)
        response.raise_for_status()

        did = response.json().get("did")
        if did:
            logger.debug(f"Resolved Bluesky handle {actor} to {did}")
        return did

    except requests.exceptions.RequestException as e:
        logger.warning(f"Error resolving Bluesky handle {actor}: {e}")
    except Exception as e:
        logger.error(f"Unexpected error resolving Bluesky handle {actor}: {e}")

    return None


def fetch_bluesky_post(actor: str, rkey: str, timeout: int = 10) -> Optional[Dict[str, Any]]:
    """
    Fetch post data from the public Bluesky API.

    Resolves the actor to a DID, constructs the AT-URI, and fetches the post
    via app.bsky.feed.getPosts.

    Args:
        actor: Bluesky handle or DID
        rkey: Post record key
        timeout: Request timeout in seconds

    Returns:
        Post data dict if successful, None if failed
    """
    if not actor or not rkey:
        return None

    did = resolve_bluesky_did(actor, timeout=timeout)
    if not did:
        return None

    at_uri = f"at://{did}/app.bsky.feed.post/{rkey}"

    try:
        url = f"{BSKY_API_BASE}/xrpc/app.bsky.feed.getPosts"
        headers = {"User-Agent": "Yana/1.0"}

        response = requests.get(url, params={"uris": at_uri}, headers=headers, timeout=timeout)
        response.raise_for_status()

        posts = response.json().get("posts") or []
        if not posts:
            logger.debug(f"Bluesky post {at_uri} not found")
            return None

        logger.debug(f"Fetched Bluesky post {at_uri}")
        return posts[0]

    except requests.exceptions.HTTPError as e:
        logger.warning(f"HTTP error fetching Bluesky post {at_uri}: {e.response.status_code}")
    except requests.exceptions.RequestException as e:
        logger.warning(f"Error fetching Bluesky post {at_uri}: {e}")
    except Exception as e:
        logger.error(f"Unexpected error fetching Bluesky post {at_uri}: {e}")

    return None


def extract_image_urls_from_post(post: Dict[str, Any]) -> List[str]:
    """
    Extract image URLs from Bluesky post data.

    Handles both app.bsky.embed.images#view and the media side of
    app.bsky.embed.recordWithMedia#view.

    Args:
        post: Post data dict from the Bluesky API

    Returns:
        List of fullsize image URLs found in the post
    """
    if not post:
        return []

    image_urls: List[str] = []

    try:
        embed = post.get("embed") or {}
        embed_type = embed.get("$type", "")

        # recordWithMedia wraps the actual media in a "media" key
        if "recordWithMedia" in embed_type:
            embed = embed.get("media") or {}

        for image in embed.get("images") or []:
            if isinstance(image, dict):
                img_url = image.get("fullsize") or image.get("thumb")
                if img_url:
                    image_urls.append(img_url)

    except (KeyError, TypeError) as e:
        logger.debug(f"Error extracting images from Bluesky post: {e}")

    return image_urls


def build_bluesky_embed_html(url: str) -> Optional[str]:
    """
    Build a rich HTML embed for a Bluesky post.

    Fetches post data from the public Bluesky API and renders it as a styled
    blockquote with author info, post text, images, and engagement stats.

    Args:
        url: Bluesky post URL

    Returns:
        HTML string with the post embed, or None if fetching failed
    """
    info = extract_bluesky_post_info(url)
    if not info:
        return None

    actor, rkey = info
    post = fetch_bluesky_post(actor, rkey)
    if not post:
        return None

    record = post.get("record") or {}
    text = record.get("text", "")
    author = post.get("author") or {}
    display_name = author.get("displayName", "")
    handle = author.get("handle", "")
    likes = post.get("likeCount", 0)
    reposts = post.get("repostCount", 0)
    replies = post.get("replyCount", 0)
    created_at = record.get("createdAt", "")

    # Clean URL (remove tracking params)
    clean_url = url.split("?")[0]

    # Build HTML parts
    parts = [
        '<blockquote style="border-left: 3px solid #0085ff; padding: 12px 16px;'
        ' margin: 1em 0; background: #f7f9fa;">',
    ]

    # Author line. clean_url is attacker-reachable (it comes from the source
    # feed/page URL) and lands inside an href attribute, so it needs both an
    # escape (quotes could break out of the attribute) and a scheme check
    # (a well-formed but unescaped `javascript:` URL is an XSS vector that
    # escaping alone does not fix -- see is_safe_url).
    author_display = display_name or (f"@{handle}" if handle else "")
    handle_suffix = f" (@{handle})" if display_name and handle else ""
    link_html = (
        f'<a href="{html.escape(clean_url, quote=True)}" target="_blank" rel="noopener">'
        f"View on Bluesky</a>"
        if is_safe_url(clean_url)
        else "View on Bluesky"
    )
    parts.append(
        f'<p style="margin: 0 0 8px 0;">'
        f"<strong>{_escape(author_display)}</strong>{_escape(handle_suffix)} · "
        f"{link_html}"
        f"</p>"
    )

    # Post text
    if text:
        parts.append(f'<p style="margin: 0 0 8px 0; white-space: pre-wrap;">{_escape(text)}</p>')

    # Images. Same reasoning as clean_url above: escape for the attribute
    # context, and skip the image entirely rather than render an unsafe
    # scheme's URL.
    for img_url in extract_image_urls_from_post(post):
        if not is_safe_url(img_url):
            continue
        parts.append(
            f'<p><img src="{html.escape(img_url, quote=True)}" alt="Bluesky image"'
            f' style="max-width: 100%; border-radius: 8px;"></p>'
        )

    # Engagement stats and date
    stats_parts = []
    if likes:
        stats_parts.append(f"&#9829; {_format_count(likes)}")
    if reposts:
        stats_parts.append(f"&#128257; {_format_count(reposts)}")
    if replies:
        stats_parts.append(f"&#128172; {_format_count(replies)}")
    if created_at:
        formatted_date = _format_post_date(created_at)
        if formatted_date:
            stats_parts.append(formatted_date)

    if stats_parts:
        stats_str = " · ".join(stats_parts)
        parts.append(f'<p style="margin: 0; color: #536471; font-size: 0.9em;">{stats_str}</p>')

    parts.append("</blockquote>")

    logger.debug(f"Built Bluesky embed for {clean_url}")
    return "\n".join(parts)


def _escape(text: str) -> str:
    """Escape HTML special characters.

    Delegates to the stdlib rather than hand-rolling the replace() chain: the
    old version covered ``&``/``<``/``>``/``"`` but missed ``'``, which is
    exploitable wherever the escaped value ends up inside a single-quoted
    attribute (or is used to close out of one). ``html.escape(..., quote=True)``
    covers all five. Same fix, same rationale, as ``twitter._escape``.
    """
    return html.escape(text, quote=True)


def _format_count(count: int) -> str:
    """Format a number for display (e.g. 1234 -> '1.2K')."""
    if count >= 1_000_000:
        return f"{count / 1_000_000:.1f}M"
    if count >= 1_000:
        return f"{count / 1_000:.1f}K"
    return str(count)


def _format_post_date(created_at: str) -> Optional[str]:
    """Format an ISO 8601 Bluesky date string for display."""
    try:
        from datetime import datetime

        # Bluesky returns ISO dates like "2026-06-04T04:34:34.364Z"
        normalized = created_at.replace("Z", "+00:00")
        dt = datetime.fromisoformat(normalized)
        return dt.strftime("%b %d, %Y")
    except (AttributeError, ValueError, TypeError):
        return None

File: src/lib/aggregators/embeds/bluesky.ts
/**
 * Bluesky post support.
 *
 * Ported from old/core/aggregators/utils/bluesky.py.
 *
 * `isBlueskyUrl`/`extractBlueskyPostInfo` recognise a Bluesky post URL and
 * pull its actor/rkey apart; `buildBlueskyEmbedHtml` fetches the post from
 * the public API and renders it as a styled blockquote. Consumed today by
 * `sites/mein_mmo/embeds.ts`, which finds a Bluesky link in a mein_mmo
 * article's own markup and asks this module to render the replacement HTML
 * directly — there is no generic embed-provider registry here (one used to
 * exist across this directory: a `detect`/`convert` pair per provider behind
 * a first-match-wins dispatch table, but every production embed actually
 * goes through `blocks/parser.ts`'s own `embedFacade()`/`tweetEmbed()`, so
 * the registry and its four provider pairs — including this module's own
 * `detectBluesky`/`convertBluesky` — had zero production callers and were
 * deleted; see `.superpowers/sdd/2026-09-03-pipeline-review-4-cleanup-and-hardening/task-1-brief.md`).
 * If a real multi-provider registry is ever wanted again, model it as a
 * declaration — `defineEmbedProvider({ key, detect, convert })` returning a
 * `{ detect, convert }` pair, in the shape `src/lib/integrations/define.ts`
 * uses for credential providers — rather than resurrecting the
 * import-side-effect registration this module used to do.
 */

import type { ChromeLabels } from "../chrome-labels";
import { escapeHtml } from "../extract/format";
import { isSafeUrl } from "../blocks/parser";

/** Public (unauthenticated) Bluesky AppView API endpoint. */
const BSKY_API_BASE = "https://public.api.bsky.app";

/** Check if a URL is a Bluesky URL. */
export function isBlueskyUrl(url: string): boolean {
  return Boolean(url) && url.includes("bsky.app");
}

/**
 * Extract the actor (handle or DID) and record key from a Bluesky post URL.
 * Pattern: /profile/{handle_or_did}/post/{rkey}
 */
export function extractBlueskyPostInfo(url: string): { actor: string; rkey: string } | null {
  if (!url) return null;
  const match = /\/profile\/([^/]+)\/post\/([^/?#]+)/.exec(url);
  return match ? { actor: match[1]!, rkey: match[2]! } : null;
}

/**
 * Accepted risk: unbounded serial network I/O under the job timeout.
 *
 * A single buildBlueskyEmbedHtml() call can take up to ~20s on transport
 * failure (resolveBlueskyDid() and fetchBlueskyPost() each carry their own
 * 10s AbortSignal.timeout, and mein_mmo's processEmbeds() awaits them one
 * figure at a time, never in parallel). A feed with many Bluesky-embedding
 * articles hit during a Bluesky outage could therefore push a single
 * aggregation run close to or past src/lib/jobs/worker.ts's 300s job-level
 * timeout (`timeoutMs`).
 *
 * This is accepted for now rather than fixed here: it matches the Python
 * origin's equally unbounded synchronous design (old/core's
 * BlueskyEmbedProcessor has no equivalent job-level wall-clock cap either),
 * Bluesky embeds are a minority embed type within any given feed's articles,
 * and a real fix -- a run-scoped circuit breaker that stops attempting
 * further Bluesky lookups for the rest of a run after the first transport
 * failure -- needs shared state threaded through
 * extractMeinMmoContent -> processEmbeds -> BlueskyEmbedProcessor, which is a
 * real architecture change deserving its own design/test cycle, not a rushed
 * addition here. Revisit if this proves problematic in practice.
 */

/**
 * Resolve a Bluesky handle to a DID.
 * If the actor is already a DID (starts with "did:"), it is returned as-is.
 */
async function resolveBlueskyDid(actor: string): Promise<string | null> {
  if (!actor) return null;
  if (actor.startsWith("did:")) return actor;

  try {
    const url = `${BSKY_API_BASE}/xrpc/com.atproto.identity.resolveHandle?handle=${encodeURIComponent(actor)}`;
    const res = await fetch(url, {
      headers: { "User-Agent": "Yana/1.0" },
      signal: AbortSignal.timeout(10_000),
    });
    if (!res.ok) return null;
    const data = (await res.json()) as { did?: string };
    return data.did ?? null;
  } catch {
    return null;
  }
}

/**
 * Fetch post data from the public Bluesky API.
 */
async function fetchBlueskyPost(
  actor: string,
  rkey: string,
): Promise<Record<string, unknown> | null> {
  const did = await resolveBlueskyDid(actor);
  if (!did) return null;

  const atUri = `at://${did}/app.bsky.feed.post/${rkey}`;
  try {
    const url = `${BSKY_API_BASE}/xrpc/app.bsky.feed.getPosts?uris=${encodeURIComponent(atUri)}`;
    const res = await fetch(url, {
      headers: { "User-Agent": "Yana/1.0" },
      signal: AbortSignal.timeout(10_000),
    });
    if (!res.ok) return null;
    const data = (await res.json()) as { posts?: Record<string, unknown>[] };
    return data.posts?.[0] ?? null;
  } catch {
    return null;
  }
}

/**
 * Extract image URLs from a Bluesky post.
 * Handles both app.bsky.embed.images#view and recordWithMedia#view.
 */
function extractImageUrls(post: Record<string, unknown>): string[] {
  const urls: string[] = [];
  try {
    let embed = (post.embed ?? {}) as Record<string, unknown>;
    const embedType = (embed.$type ?? "") as string;
    if (embedType.includes("recordWithMedia")) {
      embed = (embed.media ?? {}) as Record<string, unknown>;
    }
    const images = (embed.images ?? []) as Array<Record<string, unknown>>;
    for (const img of images) {
      const url = (img.fullsize ?? img.thumb) as string | undefined;
      if (url) urls.push(url);
    }
  } catch {
    // ignore extraction errors
  }
  return urls;
}

/** Format an engagement count for display (e.g. 1234 -> "1.2K"). */
export function formatBlueskyCount(count: number): string {
  if (count >= 1_000_000) return `${(count / 1_000_000).toFixed(1)}M`;
  if (count >= 1_000) return `${(count / 1_000).toFixed(1)}K`;
  return String(count);
}

const MONTH_ABBREVIATIONS = [
  "Jan",
  "Feb",
  "Mar",
  "Apr",
  "May",
  "Jun",
  "Jul",
  "Aug",
  "Sep",
  "Oct",
  "Nov",
  "Dec",
];

/**
 * Format an ISO 8601 Bluesky date string for display, or null if unparseable.
 * Uses UTC getters and a fixed month-name table rather than Intl: the input
 * is always UTC ("...Z"), and this avoids the display depending on the
 * server process's locale.
 */
export function formatBlueskyPostDate(createdAt: string): string | null {
  if (!createdAt) return null;
  const date = new Date(createdAt);
  if (Number.isNaN(date.getTime())) return null;
  const month = MONTH_ABBREVIATIONS[date.getUTCMonth()]!;
  const day = String(date.getUTCDate()).padStart(2, "0");
  return `${month} ${day}, ${date.getUTCFullYear()}`;
}

function stringField(obj: Record<string, unknown>, key: string): string {
  const value = obj[key];
  return typeof value === "string" ? value : "";
}

function numberField(obj: Record<string, unknown>, key: string): number {
  const value = obj[key];
  return typeof value === "number" ? value : 0;
}

/**
 * Build a rich HTML embed for a Bluesky post.
 *
 * Ported from old/core/aggregators/utils/bluesky.py's build_bluesky_embed_html.
 * Fetches post data from the public Bluesky API and renders it as a styled
 * blockquote with author info, post text, images, and engagement stats.
 *
 * Returns null when the URL isn't a post URL or the API fetch fails. The
 * caller (mein_mmo's BlueskyEmbedProcessor) removes the figure entirely in
 * that case -- matching the Python original, which never falls back to a
 * bare link.
 */
export async function buildBlueskyEmbedHtml(
  url: string,
  labels: ChromeLabels,
): Promise<string | null> {
  const info = extractBlueskyPostInfo(url);
  if (!info) return null;

  const post = await fetchBlueskyPost(info.actor, info.rkey);
  if (!post) return null;

  const record = (post.record ?? {}) as Record<string, unknown>;
  const author = (post.author ?? {}) as Record<string, unknown>;
  const text = stringField(record, "text");
  const displayName = stringField(author, "displayName");
  const handle = stringField(author, "handle");
  const likes = numberField(post, "likeCount");
  const reposts = numberField(post, "repostCount");
  const replies = numberField(post, "replyCount");
  const createdAt = stringField(record, "createdAt");

  const cleanUrl = url.split("?")[0]!;

  const parts: string[] = [
    '<blockquote style="border-left: 3px solid #0085ff; padding: 12px 16px; ' +
      'margin: 1em 0; background: #f7f9fa;">',
  ];

  // clean_url and every image URL below are attacker-reachable (they come
  // from the source page), so each needs both an escape (for the attribute
  // context) and a scheme check via isSafeUrl -- escaping alone doesn't stop
  // a well-formed but unescaped javascript: URL.
  const authorDisplay = displayName || (handle ? `@${handle}` : "");
  const handleSuffix = displayName && handle ? ` (@${handle})` : "";
  const linkHtml = isSafeUrl(cleanUrl)
    ? `<a href="${escapeHtml(cleanUrl)}" target="_blank" rel="noopener">${labels.viewOnBluesky}</a>`
    : labels.viewOnBluesky;
  parts.push(
    `<p style="margin: 0 0 8px 0;"><strong>${escapeHtml(authorDisplay)}</strong>` +
      `${escapeHtml(handleSuffix)} · ${linkHtml}</p>`,
  );

  if (text) {
    parts.push(`<p style="margin: 0 0 8px 0; white-space: pre-wrap;">${escapeHtml(text)}</p>`);
  }

  for (const imageUrl of extractImageUrls(post)) {
    if (!isSafeUrl(imageUrl)) continue;
    parts.push(
      `<p><img src="${escapeHtml(imageUrl)}" alt="Bluesky image" ` +
        `style="max-width: 100%; border-radius: 8px;"></p>`,
    );
  }

  const statsParts: string[] = [];
  if (likes) statsParts.push(`&#9829; ${formatBlueskyCount(likes)}`);
  if (reposts) statsParts.push(`&#128257; ${formatBlueskyCount(reposts)}`);
  if (replies) statsParts.push(`&#128172; ${formatBlueskyCount(replies)}`);
  const formattedDate = createdAt ? formatBlueskyPostDate(createdAt) : null;
  if (formattedDate) statsParts.push(formattedDate);

  if (statsParts.length > 0) {
    parts.push(
      `<p style="margin: 0; color: #536471; font-size: 0.9em;">${statsParts.join(" · ")}</p>`,
    );
  }

  parts.push("</blockquote>");

  return parts.join("\n");
}
